src/video/win32_v.cpp

```diff
@@ -610,32 +615,14 @@ static LRESULT CALLBACK WndProcGdi(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lP
 			_imm_props = ImmGetProperty(GetKeyboardLayout(0), IGP_PROPERTY);
 			break;
 
-		case WM_ENTERSIZEMOVE:
-			in_sizemove = true;
-			break;
-
-		case WM_EXITSIZEMOVE:
-			in_sizemove = false;
-			break;
-
-		case WM_PAINT:
-			if (!in_sizemove && _draw_mutex != nullptr && !HasModalProgress()) {
-				/* Get the union of the old update rect and the new update rect. */
-				RECT r;
-				GetUpdateRect(hwnd, &r, FALSE);
-				UnionRect(&_wnd.update_rect, &_wnd.update_rect, &r);
-
-				/* Mark the window as updated, otherwise Windows would send more WM_PAINT messages. */
-				ValidateRect(hwnd, nullptr);
-				_draw_signal->notify_one();
-			} else {
-				PAINTSTRUCT ps;
+		case WM_PAINT: {
+			RECT r;
+			GetUpdateRect(hwnd, &r, FALSE);
+			static_cast<VideoDriver_Win32 *>(VideoDriver::GetInstance())->MakeDirty(r.left, r.top, r.right - r.left, r.bottom - r.top);
```

Comment:
I believe you're required to always call `BeginPaint` and `EndPaint` from WM_PAINT handler, even if you have nothing to do.

Reply:
`Upon receiving a WM_PAINT message, an application can call BeginPaint to retrieve the display device context for the client area`

https://docs.microsoft.com/en-us/windows/win32/gdi/the-wm-paint-message

I read that as, it is not required. The old code also didn't do it when threaded, so I assumed it was fine to do. But, this is all based on assumptions .. "it works"? :D

Reply:
https://docs.microsoft.com/en-us/windows/win32/gdi/wm-paint
> An application may call the GetUpdateRect function to determine whether the window has an update region. If GetUpdateRect returns zero, the application need not call the BeginPaint and EndPaint functions.
I read that as being required to call them, unless the update rect is empty.

Reply:
Documentation being confusing? NEVAH!

Anyway, as mentioned, the old code already wasn't doing this, so if it was broken, it is still broken. Doesn't mean we shouldn't look into it, but at least I don't break it any further, I think :D

I will add some `DEBUG` statements to see what `WM_PAINT` is doing for us.

Reply:
I think the answer is this:

> BeginPaint sets the update region of a window to NULL. This clears the region, preventing it from generating subsequent WM_PAINT messages. If an application processes a WM_PAINT message but does not call BeginPaint or otherwise clear the update region, the application continues to receive WM_PAINT messages as long as the region is not empty. In all cases, an application must clear the update region before returning from the WM_PAINT message.

Basically, call `ValidateRect` to validate the entire window, and you don't need to call `BeginPaint`.

Reply:
So I am doing the right thing by accident, whoho! :D

Reply:
Fun fact: `WM_PAINT` is only called once during the running of OpenTTD now, and once every time you unminimize the window or maximize it. Very rare event now :D